File: modules/vm-webrtc/src/ToolkitHelper.ts
import { log } from '../../../lib/logger';
import { type ToolNativeModule } from './ToolHelper';
import { executeToolkitFunction } from './toolkit_functions/toolkit_functions';
import {
  type ToolSessionContext,
  isToolSessionContextEmpty,
  summarizeToolSessionContext,
} from './toolkit_functions/types';

// MARK: - Types

export interface ToolkitRequestEvent {
  requestId: string;
  callId: string;
  groupName: string;
  toolName: string;
  arguments: string;
  eventId?: string;
}

export interface ToolkitHelperNativeModule extends ToolNativeModule {
  sendToolkitResponse(requestId: string, result: string): void;
}

// MARK: - Toolkit Helper Manager

/**
 * Manages Gen2 toolkit tool calls between JavaScript and native Swift code.
 * Uses a mux/demux approach to handle all toolkit tools through a single delegate.
 * Also maintains per-tool session context for stateful tool interactions.
 */
export class ToolkitHelper {
  private readonly toolName = 'ToolkitHelper';
  private readonly requestEventName = 'onToolkitRequest';
  private readonly module: ToolkitHelperNativeModule | null;
  private readonly toolSessionContexts: Map<string, ToolSessionContext> = new Map();

  constructor(nativeModule: ToolkitHelperNativeModule | null) {
    this.module = nativeModule;

    if (this.module) {
      this.module.addListener(this.requestEventName, this.handleRequest.bind(this));
      log.info('[ToolkitHelper] Initialized with native module', {}, { eventName: this.requestEventName });
    } else {
      log.warn('[ToolkitHelper] Native module unavailable', {});
    }
  }

  // MARK: - Private Methods

  /**
   * Handle a toolkit request from Swift.
   */
  private async handleRequest(event: ToolkitRequestEvent) {
    const { requestId, callId, groupName, toolName, arguments: argumentsJSON, eventId } = event;
    log.info(`[${this.toolName}] 📥 Received toolkit request from Swift`, {}, {
      requestId,
      callId,
      groupName,
      toolName,
      eventId,
      argumentsLength: argumentsJSON?.length || 0,
      arguments: argumentsJSON,
    });

    try {
      // Parse arguments
      let args: any = {};
      if (argumentsJSON) {
        try {
          args = JSON.parse(argumentsJSON);
        } catch (parseError) {
          log.warn(`[${this.toolName}] Failed to parse arguments JSON`, {}, {
            requestId,
            argumentsJSON,
            error: parseError instanceof Error ? parseError.message : String(parseError),
          });
        }
      }

      // Execute the toolkit operation (stubbed for now)
      const result = await this.executeToolkitOperation(groupName, toolName, args, { requestId, callId, eventId });

      log.info(`[${this.toolName}] ✅ Toolkit operation completed`, {}, {
        requestId,
        callId,
        groupName,
        toolName,
        resultLength: String(result).length,
        result: result,
      });

      if (this.module) {
      this.module.sendToolkitResponse(requestId, result);
      log.info(`[${this.toolName}] 📤 Sent response to Swift`, {}, {
        requestId,
        callId,
        responseLength: String(result).length,
        response: result,
      });
      } else {
        log.warn(`[${this.toolName}] ⚠️ Cannot send response; native module missing`, {}, { requestId });
      }
    } catch (error) {
      log.error(`[${this.toolName}] ❌ Toolkit operation failed`, {}, {
        requestId,
        callId,
        groupName,
        toolName,
        eventId,
        errorMessage: error instanceof Error ? error.message : String(error),
        errorStack: error instanceof Error ? error.stack : undefined,
        errorName: error instanceof Error ? error.name : undefined,
      }, error);

      if (this.module) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        const errorResult = JSON.stringify({ error: errorMessage });
        this.module.sendToolkitResponse(requestId, errorResult);
        log.info(`[${this.toolName}] 📤 Sent error response to Swift`, {}, {
          requestId,
          callId,
          errorMessage,
        });
      }
    }
  }

  /**
   * Execute a toolkit operation by routing to the appropriate toolkit function.
   * Handles per-tool session context round-tripping.
   * TODO: can this be DRY'd with ToolManager.executeGen2ToolCall()?
   */
  private async executeToolkitOperation(
    groupName: string,
    toolName: string,
    args: any,
    context: { requestId: string; callId: string; eventId?: string }
  ): Promise<string> {
    const { requestId, callId, eventId } = context;

    log.info(`[${this.toolName}] 🔧 Executing toolkit operation`, {}, {
      groupName,
      toolName,
      args,
      requestId,
      callId,
      eventId,
    });

    try {
      // Prepare context parameters based on the toolkit group
      let context_params;
      if (groupName === 'web') {
        context_params = {
          maxLength: 1500,
          minHtmlForBody: 15000,
          maxRawBytes: 3000000,
        };
      }

      // Get the current session context for this tool
      const toolKey = `${groupName}__${toolName}`;
      const currentSessionContext = this.toolSessionContexts.get(toolKey) || {};

      // Route to the appropriate toolkit function
      const toolkitResult = await executeToolkitFunction(
        groupName,
        toolName,
        args,
        context_params,
        currentSessionContext
      );

      // Store the updated session context for this tool
      this.toolSessionContexts.set(toolKey, toolkitResult.updatedToolSessionContext);

      // Append session context to result if non-empty
      let finalResult = toolkitResult.result;
      if (!isToolSessionContextEmpty(toolkitResult.updatedToolSessionContext)) {
        const contextSummary = summarizeToolSessionContext(toolkitResult.updatedToolSessionContext);
        finalResult = `${toolkitResult.result}\n\nTool session context: ${contextSummary}`;
      }

      log.info(`[${this.toolName}] ✅ Toolkit function executed successfully`, {}, {
        groupName,
        toolName,
        requestId,
        callId,
        resultLength: finalResult.length,
        result: finalResult,
        sessionContextKeys: Object.keys(toolkitResult.updatedToolSessionContext),
      });

      return finalResult;
    } catch (error) {
      log.error(`[${this.toolName}] ❌ Toolkit function execution failed`, {}, {
        groupName,
        toolName,
        requestId,
        callId,
        errorMessage: error instanceof Error ? error.message : String(error),
        errorStack: error instanceof Error ? error.stack : undefined,
      }, error);

      // Return error as JSON
      return JSON.stringify({
        success: false,
        error: error instanceof Error ? error.message : String(error),
        metadata: {
          groupName,
          toolName,
          timestamp: new Date().toISOString(),
        }
      });
    }
  }

  // MARK: - Public Methods

  /**
   * Execute a toolkit operation directly (for testing).
   */
  async execute(groupName: string, toolName: string, args: any): Promise<string> {
    return this.executeToolkitOperation(groupName, toolName, args, {
      requestId: 'direct-call',
      callId: 'direct-call',
    });
  }
}

// MARK: - Factory Function

/**
 * Creates a new ToolkitHelper instance with the provided native module.
 * Returns null if the module is not available.
 */
export const createToolkitHelper = (nativeModule: ToolkitHelperNativeModule | null): ToolkitHelper | null => {
  if (!nativeModule) {
    log.warn('[ToolkitHelper] Native module not available, toolkit helper will not be initialized', {});
    return null;
  }
  return new ToolkitHelper(nativeModule);
};

File: modules/vm-webrtc/ios/ToolkitHelper.swift
import Foundation
import ExpoModulesCore

// MARK: - Toolkit Tool Manager

/// Manages Gen2 toolkit tool calls between OpenAI WebRTC and JavaScript
/// Handles all toolkit-based tools using a mux/demux approach with a single delegate
public class ToolkitHelper: BaseTool {

  // MARK: - Properties

  public let toolName = "toolkit_helper"

  private weak var module: Module?
  private weak var responder: ToolCallResponder?
  private let helper: ToolHelper
  private let logger = VmWebrtcLogging.logger

  // Track pending callbacks by requestId
  private var stringCallbacks: [String: (String?, Error?) -> Void] = [:]

  // Track callback usage count to detect anomalies
  private var callbackUsageCount: [String: Int] = [:]

  // MARK: - Initialization

  public init(module: Module, responder: ToolCallResponder) {
    self.module = module
    self.responder = responder
    self.helper = ToolHelper(module: module)
    self.logger.log("[ToolkitHelper] Initialized")
  }

  // MARK: - Public Methods

  /// Handle a toolkit tool call from OpenAI
  /// - Parameters:
  ///   - callId: The tool call identifier from OpenAI
  ///   - toolName: The fully qualified tool name (e.g., "hacker_news__showTopStories")
  ///   - argumentsJSON: JSON string containing the tool arguments
  public func handleToolkitCall(callId: String, toolName: String, argumentsJSON: String) {
    self.logger.log(
      "[ToolkitHelper] Processing toolkit tool call",
      attributes: [
        "callId": callId,
        "toolName": toolName,
        "arguments_length": argumentsJSON.count,
        "arguments": argumentsJSON
      ]
    )

    // Parse the tool name to extract group and tool name
    guard let (groupName, shortToolName) = parseToolName(toolName) else {
      self.logger.log(
        "[ToolkitHelper] Failed to parse tool name",
        attributes: [
          "callId": callId,
          "toolName": toolName
        ]
      )
      responder?.sendToolCallError(callId: callId, error: "Invalid toolkit tool name format: \(toolName)")
      return
    }

    self.logger.log(
      "[ToolkitHelper] Parsed toolkit tool name",
      attributes: [
        "callId": callId,
        "groupName": groupName,
        "toolName": shortToolName
      ]
    )

    // Execute the toolkit operation
    executeToolkitOperation(
      callId: callId,
      groupName: groupName,
      toolName: shortToolName,
      argumentsJSON: argumentsJSON
    )
  }

  /// Handle a toolkit tool call from OpenAI (BaseTool protocol requirement)
  /// - Parameters:
  ///   - callId: The tool call identifier
  ///   - argumentsJSON: JSON string containing the arguments
  public func handleToolCall(callId: String, argumentsJSON: String) {
    // This is called from the event handler but we need the tool name
    // which should be passed separately. Log a warning for now.
    self.logger.log(
      "[ToolkitHelper] handleToolCall called without tool name",
      attributes: [
        "callId": callId,
        "note": "Use handleToolkitCall with explicit toolName instead"
      ]
    )
    responder?.sendToolCallError(callId: callId, error: "Toolkit tool called without tool name")
  }

  /// Handle a response from JavaScript
  /// - Parameters:
  ///   - requestId: The unique request identifier
  ///   - result: The response string from JavaScript
  public func handleResponse(requestId: String, result: String) {
    self.logger.log(
      "[ToolkitHelper] 📥 Received toolkit response from JavaScript",
      attributes: [
        "requestId": requestId,
        "result_preview": String(result.prefix(500)),
        "result": result
      ]
    )

    // Track usage count for anomaly detection
    let currentUsageCount = callbackUsageCount[requestId] ?? 0
    let newUsageCount = currentUsageCount + 1
    callbackUsageCount[requestId] = newUsageCount

    // Check for anomalous behavior (callback used more than once)
    if newUsageCount > 1 {
      self.logger.log(
        "[ToolkitHelper] 🚨 ANOMALY DETECTED: Callback used multiple times!",
        attributes: [
          "requestId": requestId,
          "usageCount": newUsageCount,
          "warning": "This indicates a bug - requestId may be reused or callback invoked multiple times"
        ]
      )
    }

    if let callback = stringCallbacks[requestId] {
      callback(result, nil)
      // Don't remove the callback - keep it to track usage patterns
      self.logger.log(
        "[ToolkitHelper] ✅ Toolkit callback executed successfully",
        attributes: [
          "requestId": requestId,
          "result_length": result.count,
          "usageCount": newUsageCount,
          "isFirstUse": newUsageCount == 1
        ]
      )
    } else {
      self.logger.log(
        "[ToolkitHelper] ⚠️ No callback found for requestId",
        attributes: [
          "requestId": requestId,
          "usageCount": newUsageCount,
          "note": "Callback may have been cleaned up or never registered"
        ]
      )
    }
  }

  /// Handle a response from JavaScript (BaseTool protocol requirement for Int results)
  /// - Parameters:
  ///   - requestId: The unique request identifier
  ///   - result: The integer result
  public func handleResponse(requestId: String, result: Int) {
    self.logger.log(
      "[ToolkitHelper] ⚠️ Received int result, but toolkit expects string",
      attributes: [
        "requestId": requestId,
        "result": result
      ]
    )
  }

  // MARK: - Private Methods

  /// Parse a fully qualified tool name into group and tool name
  /// - Parameter fullName: The full tool name (e.g., "hacker_news__showTopStories")
  /// - Returns: Tuple of (groupName, toolName) or nil if invalid format
  private func parseToolName(_ fullName: String) -> (String, String)? {
    // Look for the double underscore separator "__"
    guard let range = fullName.range(of: "__") else {
      self.logger.log(
        "[ToolkitHelper] Tool name does not contain double underscore separator",
        attributes: ["fullName": fullName]
      )
      return nil
    }

    let groupName = String(fullName[..<range.lowerBound])
    let toolName = String(fullName[range.upperBound...])

    // Validate both parts are non-empty
    guard !groupName.isEmpty, !toolName.isEmpty else {
      self.logger.log(
        "[ToolkitHelper] Tool name has empty group or tool component",
        attributes: [
          "fullName": fullName,
          "groupName": groupName,
          "toolName": toolName
        ]
      )
      return nil
    }

    return (groupName, toolName)
  }

  /// Execute a toolkit operation by forwarding to JavaScript
  /// - Parameters:
  ///   - callId: The OpenAI tool call identifier
  ///   - groupName: The toolkit group name (e.g., "hacker_news")
  ///   - toolName: The tool name (e.g., "showTopStories")
  ///   - argumentsJSON: JSON string containing the arguments
  private func executeToolkitOperation(
    callId: String,
    groupName: String,
    toolName: String,
    argumentsJSON: String
  ) {
    self.logger.log(
      "[ToolkitHelper] Executing toolkit tool call",
      attributes: [
        "callId": callId,
        "groupName": groupName,
        "toolName": toolName,
        "arguments_length": argumentsJSON.count
      ]
    )

    // Generate a request ID for tracking this operation
    let requestId = ToolHelper.generateRequestId()

    self.logger.log(
      "[ToolkitHelper] Generated request ID for toolkit operation",
      attributes: [
        "callId": callId,
        "requestId": requestId,
        "groupName": groupName,
        "toolName": toolName
      ]
    )

    // Register callback for this request
    registerStringCallback(requestId: requestId) { result, error in
      if let error = error {
        self.logger.log(
          "[ToolkitHelper] Toolkit operation failed",
          attributes: [
            "callId": callId,
            "requestId": requestId,
            "groupName": groupName,
            "toolName": toolName,
            "error": error.localizedDescription
          ]
        )
        self.responder?.sendToolCallError(callId: callId, error: error.localizedDescription)
        return
      }

      guard let result = result else {
        self.logger.log(
          "[ToolkitHelper] Toolkit operation returned no result",
          attributes: [
            "callId": callId,
            "requestId": requestId,
            "groupName": groupName,
            "toolName": toolName
          ]
        )
        self.responder?.sendToolCallError(callId: callId, error: "No result from toolkit operation")
        return
      }

      self.logger.log(
        "[ToolkitHelper] Toolkit operation result received",
        attributes: [
          "callId": callId,
          "requestId": requestId,
          "groupName": groupName,
          "toolName": toolName,
          "result_preview": String(result.prefix(1000))
        ]
      )

      // Send the result back to OpenAI
      self.responder?.sendToolCallResult(callId: callId, result: result)
    }

    // Emit event to JavaScript
    self.logger.log(
      "[ToolkitHelper] 📤 Emitting toolkit request to JavaScript",
      attributes: [
        "eventName": "onToolkitRequest",
        "requestId": requestId,
        "groupName": groupName,
        "toolName": toolName,
        "callId": callId,
        "arguments": argumentsJSON
      ]
    )

    let eventId = helper.emitToolRequest(
      eventName: "onToolkitRequest",
      requestId: requestId,
      parameters: [
        "callId": callId,
        "groupName": groupName,
        "toolName": toolName,
        "arguments": argumentsJSON
      ]
    )

    self.logger.log(
      "[ToolkitHelper] 🆔 Event emitted to JavaScript",
      attributes: [
        "requestId": requestId,
        "eventId": eventId,
        "groupName": groupName,
        "toolName": toolName,
        "arguments": argumentsJSON
      ]
    )

    // Set up timeout
    setupStringTimeout(for: requestId, errorMessage: "Toolkit operation timed out")
  }

  /// Register a callback for a string result
  /// - Parameters:
  ///   - requestId: The unique request identifier
  ///   - callback: Callback to invoke when response is received
  private func registerStringCallback(requestId: String, callback: @escaping (String?, Error?) -> Void) {
    stringCallbacks[requestId] = callback
    self.logger.log(
      "[ToolkitHelper] 🔐 Registered string callback",
      attributes: [
        "requestId": requestId,
        "pendingCallbacks": stringCallbacks.count
      ]
    )
  }

  /// Set up a timeout for a toolkit operation
  /// - Parameters:
  ///   - requestId: The unique request identifier
  ///   - errorMessage: Error message to use if timeout occurs
  private func setupStringTimeout(for requestId: String, errorMessage: String) {
    self.logger.log(
      "[ToolkitHelper] ⏱️ Scheduling timeout",
      attributes: [
        "requestId": requestId,
        "timeoutSeconds": 60
      ]
    )

    DispatchQueue.main.asyncAfter(deadline: .now() + 60.0) { [weak self] in
      guard let self = self else { return }

      // Check if callback has already been used
      let usageCount = self.callbackUsageCount[requestId] ?? 0

      if let callback = self.stringCallbacks[requestId], usageCount == 0 {
        self.logger.log(
          "[ToolkitHelper] ⏰ Request timed out",
          attributes: [
            "requestId": requestId,
            "usageCount": usageCount
          ]
        )
        let error = NSError(
          domain: "ToolkitHelper",
          code: -1,
          userInfo: [NSLocalizedDescriptionKey: errorMessage]
        )

        // Mark as used before calling callback
        self.callbackUsageCount[requestId] = 1

        callback(nil, error)
        // Don't remove the callback - keep it to track usage patterns
      } else if usageCount > 0 {
        self.logger.log(
          "[ToolkitHelper] ⏰ Timeout fired but callback already used",
          attributes: [
            "requestId": requestId,
            "usageCount": usageCount
          ]
        )
      }
    }
  }
}
